lib/unicode.js

```diff
@@ -0,0 +1,210 @@
+'use strict';
+
+const has_icu = process.binding('config').hasIntl;
+const Buffer = require('buffer').Buffer;
+const kDefaultOptions = {};
+
+if (has_icu) {
+  const icu = process.binding('icu');
+
+  const conversions = {
+    'ascii': {
+      'binary': (source) => {
+        return source;
+      },
+      'utf8': (source) => {
+        return source;
+      },
+      'ucs2': (source) => {
+        return icu.convertToUcs2('us-ascii', source);
+      }
+    },
+    'binary': {
+      'ascii': (source) => {
+        return icu.convert('us-ascii', 'iso8859-1', source);
+      },
+      'utf8': (source) => {
+        return icu.convert('utf-8', 'iso8859-1', source);
+      },
+      'ucs2': (source) => {
+        return icu.convertToUcs2('iso8859-1', source);
+      }
+    },
+    'utf8': {
+      'ascii': (source) => {
+        return icu.convert('us-ascii', 'utf-8', source);
+      },
+      'binary': (source) => {
+        return icu.convert('iso-8859-1', 'utf-8', source);
+      },
+      'ucs2': icu.convertToUcs2FromUtf8,
+    },
+    'ucs2': {
+      'ascii': (source) => {
+        if (source.length % 2 !== 0)
+          throw new TypeError('Invalid UCS2 Buffer');
+        return icu.convertFromUcs2('us-ascii', source);
+      },
+      'binary': (source) => {
+        if (source.length % 2 !== 0)
+          throw new TypeError('Invalid UCS2 Buffer');
+        return icu.convertFromUcs2('iso-8859-1', source);
+      },
+      'utf8': (source) => {
+        if (source.length % 2 !== 0)
+          throw new TypeError('Invalid UCS2 Buffer');
+        return icu.convertToUtf8FromUcs2(source);
+      }
+    }
+  };
+
+  exports.transcode = function transcode(source, from_enc, to_enc, options) {
+    if (!source) return;
+    if (!(source.buffer instanceof ArrayBuffer))
+      throw new TypeError('"source" argument must be a Buffer');
+    if (source.length === 0) return Buffer.alloc(0);
+
+    from_enc = normalizeEncoding(from_enc);
+    to_enc = normalizeEncoding(to_enc);
+
+    if (from_enc === to_enc)
+      return source;
+
+    options = options || kDefaultOptions;
+    const cnv_from = conversions[from_enc];
+
+    if (cnv_from) {
+      const cnv_to = cnv_from[to_enc];
+      if (cnv_to)
+        return cnv_to(source, options.lenient);
+    }
+    throw new Error(`Unsupported conversion: ${from_enc} to ${to_enc}`);
+  };
+
+  function normalizeEncoding(enc) {
+    if (!enc) return 'utf8';
+    enc = String(enc).toLowerCase();
+    switch (enc) {
+      case 'utf-8':
+        return 'utf8';
+      case 'us-ascii':
+        return 'ascii';
+      case 'iso-8859-1':
+      case 'latin-1':
+      case 'latin1':
+      case 'binary':
+        return 'binary';
+      case 'ucs-2':
+      case 'utf16le':
+      case 'utf-16le':
+      case 'utf16-le':
+        return 'ucs2';
+      default:
+        return enc;
```

Comment:
Propose to throw instead of returning, throwing might be more easy to notify the wrong inputs from user-land.

```js
default: 
  throw new TypeError(`encoding ${enc} is invalid`);
```